### DISCUSSION TITLE
v1.1.0-alpha.1 - Experimental Mixture-of-Experts (MoE) Release

### DISCUSSION THREAD

**Discussion by MeryylleA** [MeryylleA/lunariscodex#114] [Ideas]
This is the first **alpha** and **experimental** release introducing the **Mixture-of-Experts (MoE)** architecture to Lunaris Codex. This release is intended for testing, community feedback, and research. It is not recommended for production use.

### ✨ New Features & Major Changes

*   **MoE Architecture (`model_moe.py`):**
    *   Implementation of a Switch Transformer-style (`k=1`) `MixtureOfExperts` layer, which replaces the standard FFN in the Transformer blocks.
    *   Added the calculation of an **auxiliary load-balancing loss** to ensure uniform expert utilization and prevent training collapse.
    *   The model's `forward` method now returns a loss tuple `(total, main, aux)` to allow for detailed monitoring.
*   **Adapted Training Script (`train_moe.py`):**
    *   A new, dedicated training script that handles the MoE model's new loss structure.
    *   Enhanced logging to both the console and Weights & Biases, tracking `main_loss`, `aux_loss`, and `total_loss` separately.
    *   Perplexity is now correctly calculated using only the `main_loss` (cross-entropy).

### 📝 Documentation

*   **Updated README:** A new `README.md` specific to the `moe-experiment` branch has been created, detailing the new architecture, training pipeline changes, and how to configure and run an MoE training job.
*   **Code Comments:** The `model_moe.py` file has been extensively commented to explain the logic behind the MoE implementation, the routing mechanics, and the auxiliary loss calculation.

### ⚠️ Known Issues & Limitations

*   **Routing Performance:** The token routing to experts is implemented with a simple Python loop. While correct and readable, it is not optimized for maximum throughput. Future releases may explore custom CUDA kernels.
*   **Memory Usage:** The model's total parameter count (and thus its memory footprint for storage) scales linearly with the number of experts.

---

To get started, check out the `moe-experiment` branch and follow the instructions in the new `README.md`.

```bash
git checkout moe-experiment
```

We appreciate any feedback and look forward to seeing the community's results

<hr /><em>This discussion was created from the release <a href='https://github.com/MeryylleA/lunariscodex/releases/tag/v1.1.0-alpha.1'>v1.1.0-alpha.1 - Experimental Mixture-of-Experts (MoE) Release</a>.</em>
